### PR TITLE
media parameter attribute value needs to escape quotes inside quoted …

### PR DESCRIPTION
**Pull request by requaos**
…value

### CODE REVIEW

File: header.go [hunk 1]
```diff
@@ -513,6 +514,85 @@ func fixUnquotedSpecials(s string) string {
 	return clean.String()
 }
```
Comment (jhillyerd) [inline header.go]: Could you add a very short description of func with an example so the reader can visualize the transformation it's making?

File: header.go [hunk 2]
```diff
@@ -513,6 +514,85 @@ func fixUnquotedSpecials(s string) string {
...
 
+func fixUnescapedQuotes(s string) string {
+	split := strings.SplitAfter(s, ";")
+	fresh := &strings.Builder{}
```
Comment (jhillyerd) [inline header.go]:
```suggestion
	sb:= &strings.Builder{}
```

for String Builder; although `fresh` did make me smile.

File: header.go [hunk 3]
```diff
@@ -513,6 +514,85 @@ func fixUnquotedSpecials(s string) string {
...
+	for i := 0; i < len(split); i++ {
+		// find the "="
+		idx := strings.IndexByte(split[i], '=')
+		if idx < 0 {
+			// no "=", must be the content-type or a comment
```
Comment (jhillyerd) [inline header.go]:
```suggestion
			// No "=", must be the content-type or a comment.
```

File: header.go [hunk 4]
```diff
@@ -513,6 +514,85 @@ func fixUnquotedSpecials(s string) string {
 	return clean.String()
 }
 
+func fixUnescapedQuotes(s string) string {
+	split := strings.SplitAfter(s, ";")
+	fresh := &strings.Builder{}
+	for i := 0; i < len(split); i++ {
+		// find the "="
+		idx := strings.IndexByte(split[i], '=')
```
Comment (jhillyerd) [inline header.go]:
```suggestion
		eq := strings.IndexByte(split[i], '=')
```

I don't see `idx` used in Go; for example, it's not used once in the entire strings pkg.  I do think that the Go authors sometimes overuse the single character variables which can be confusing in longer functions.

File: header.go [hunk 5]
```diff
@@ -513,6 +514,85 @@ func fixUnquotedSpecials(s string) string {
...
+		if idx < 0 {
+			// no "=", must be the content-type or a comment
+			fresh.WriteString(split[i])
+			continue
+		}
+		fresh.WriteString(split[i][:idx])
+		param := split[i][idx:]
+		openingQuoteIdx := strings.IndexByte(param, '"')
```
Comment (jhillyerd) [inline header.go]:
```suggestion
		openingQuote := strings.IndexByte(param, '"')
```

File: header.go [hunk 6]
```diff
@@ -513,6 +514,85 @@ func fixUnquotedSpecials(s string) string {
...
+		closingQuoteIdx := strings.LastIndexByte(param, '"')
```
Comment (jhillyerd) [inline header.go]:
```suggestion
		closingQuote := strings.LastIndexByte(param, '"')
```

File: header.go [hunk 7]
```diff
@@ -513,6 +514,85 @@ func fixUnquotedSpecials(s string) string {
...
+		// cut out early if there are no quotes
```
Comment (jhillyerd) [inline header.go]:
```suggestion
		// Cut out early if there are no quotes.
```

File: header.go [hunk 8]
```diff
@@ -513,6 +514,85 @@ func fixUnquotedSpecials(s string) string {
...
+		if openingQuoteIdx < 0 && closingQuoteIdx < 0 {
+			// value isn't quoted
```
Comment (jhillyerd) [inline header.go]:
```suggestion
			// Value isn't quoted.
```

File: header.go [hunk 9]
```diff
@@ -513,6 +514,85 @@ func fixUnquotedSpecials(s string) string {
...
+			fresh.WriteString(param)
+			continue
+		}
+		// lets append the next chunk of split here in case of a semicolon mid string
```
Comment (jhillyerd) [inline header.go]:
```suggestion
		// Lets append the next chunk of split here in case of a semicolon mid string.
```

File: header.go [hunk 10]
```diff
@@ -513,6 +514,85 @@ func fixUnquotedSpecials(s string) string {
...
+		if closingQuoteIdx == openingQuoteIdx {
+			param = fmt.Sprintf("%s%s", param, split[i+1])
+			closingQuoteIdx = strings.LastIndexByte(param, '"')
+			i++
+			if closingQuoteIdx == openingQuoteIdx {
+				return fresh.String()
+			}
+		}
+		// it's got quotes, lets put the k/v separator back in along with everything upto the first quote
```
Comment (jhillyerd) [inline header.go]:
```suggestion
		// It's got quotes, lets put the k/v separator back in along with everything up to the first quote.
```

File: header.go [hunk 11]
```diff
@@ -513,6 +514,85 @@ func fixUnquotedSpecials(s string) string {
...
+		fresh.WriteByte('=')
+		// quotes start here
```
Comment (jhillyerd) [inline header.go]:
```suggestion
		// Quotes start here.
```

File: header.go [hunk 12]
```diff
@@ -513,6 +514,85 @@ func fixUnquotedSpecials(s string) string {
...
+		fresh.WriteByte('"')
+		fresh.WriteString(param[1:openingQuoteIdx])
+		// get just the value, less the outer quotes
+		rest := param[closingQuoteIdx+1:]
+		// if there is stuff after the last quote then we should escape the first quote
```
Comment (jhillyerd) [inline header.go]:
```suggestion
		// If there is stuff after the last quote then we should escape the first quote.
```

File: header.go [hunk 13]
```diff
@@ -513,6 +514,85 @@ func fixUnquotedSpecials(s string) string {
...
+		if len(rest) > 0 && rest != ";" {
+			fresh.WriteString("\\\"")
+		}
+		param = param[openingQuoteIdx+1 : closingQuoteIdx]
+		alreadyEscaped := false
+		for strIdx := range param {
```
Comment (jhillyerd) [inline header.go]: I don't feel like the name `strIdx` conveys anything unique, and this loop is fairly short.

```suggestion
		for j := range param {
```

File: header.go [hunk 14]
```diff
@@ -513,6 +514,85 @@ func fixUnquotedSpecials(s string) string {
...
+			switch param[strIdx] {
+			case '"':
+				// we are inside of a quoted string, so lets escape this guy if it isn't already escaped
```
Comment (jhillyerd) [inline header.go]:
```suggestion
				// We are inside of a quoted string, so lets escape this guy if it isn't already escaped.
```

File: header.go [hunk 15]
```diff
@@ -513,6 +514,85 @@ func fixUnquotedSpecials(s string) string {
...
+				if !alreadyEscaped {
+					fresh.WriteByte('\\')
+					alreadyEscaped = false
+				}
+				fresh.WriteByte(param[strIdx])
+			case '\\':
+				// something is getting escaped, a quote is the only char that needs
```
Comment (jhillyerd) [inline header.go]:
```suggestion
				// Something is getting escaped, a quote is the only char that needs
```

File: header.go [hunk 16]
```diff
@@ -513,6 +514,85 @@ func fixUnquotedSpecials(s string) string {
...
+				// this, so lets assume the following char is a double-quote
```
Comment (jhillyerd) [inline header.go]:
```suggestion
				// this, so lets assume the following char is a double-quote.
```

File: header.go [hunk 17]
```diff
@@ -513,6 +514,85 @@ func fixUnquotedSpecials(s string) string {
...
+				alreadyEscaped = true
+				fresh.WriteByte('\\')
+			default:
+				fresh.WriteByte(param[strIdx])
+			}
+		}
+		// if there is stuff after the last quote then we should escape
```
Comment (jhillyerd) [inline header.go]:
```suggestion
		// If there is stuff after the last quote then we should escape
```

File: header.go [hunk 18]
```diff
@@ -513,6 +514,85 @@ func fixUnquotedSpecials(s string) string {
...
+		// the last quote, apply the rest and terminate with a quote
```
Comment (jhillyerd) [inline header.go]:
```suggestion
		// the last quote, apply the rest and terminate with a quote.
```

File: header.go (hunk 1 above)
Comment (jhillyerd) [reply header.go]: I think GH got confused here, this was meant to go above the func fixUnescapedQuotes definition

File: header.go [hunk 19]
```diff
@@ -513,6 +514,82 @@ func fixUnquotedSpecials(s string) string {
 	return clean.String()
 }
 
+func fixUnescapedQuotes(s string) string {
```
[the remark this answers is not in the document]
Comment (requaos) [reply header.go]: No, I totally agree. When I am visualizing how a stream is being read or how functions will interact, I just throw var names at the wall. Going back over it once functionality is achieved to apply Go Idioms is just a part of the development process. Thank you.

File: header.go (hunk 4 above)
Comment (requaos) [reply header.go]: Huh, I must have carried that over from my C# or Java days, lol. Good point. I believe that shorter var names would make the intermediate AST smaller during compile time, but in practice I think it makes it easier to see the business logic abstracted away from the preconceptions that wordy var names can imply... I don't practice this enough, so I digress.

**Comment by requaos** [comment]
Thank you for the feedback. I have applied your suggestions where appropriate and some I just reworded (with capitalization and punctuation in mind). And I fixed a bug related to `escaped` state where if it was not a `"` character following the `\` the flag would continue to indicate that an escape character had preceded it when it eventually encounters a `"` character.

**Comment by requaos** [comment]
Please let me know if you have any additional feedback

**Pull request merged**